lib/StaticUniformGrid.js

```diff
@@ -0,0 +1,277 @@
+ *                   `min` and `max` in place.
+ * @param {Function} cellCheckFunction A function for checking if an object intersects a grid cell, given as 2 Cartesian3s.
+ *                   This function should expect to take an object and the bounding cell's `min` and cell width. It should
+ *                   return `true` if the object intersects with the cell at all and `false` otherwise.
+ * @param {Number} cellWidth, The cell width of the uniform grid.
+ */
+
+var minScratch = new Cartesian3();
+var maxScratch = new Cartesian3();
+
+function StaticUniformGrid(data, cellWidth, objectAABBFunction, cellCheckFunction) {
+
+    /**
+     * The width of a single cell.
+     * @type {Number}
+     */
+    this.cellWidth = cellWidth;
+
+    /**
+     * The axis aligned bounding box of the grid
+     * @type {AxisAlignedBoundingBox}
+     */
+    this.AABB = new AxisAlignedBoundingBox();
+
+    /**
+     * The cell count in each dimension.
+     * @type {Cartesian3}
+     */
+    this.resolution = new Cartesian3();
+
+    /**
+     * Start index for each cell's data within the data array.
+     * @type {Number[]}
+     */
+    this.cellIndices = [];
+
+    /**
+     * Count for each cell's data within the data array.
+     * @type {Number[]}
+     */
+    this.cellCounts = [];
+
+    /**
+     * Array of objects
+     * @type {Array}
+     */
+    this.data = [];
+
+    ////// Populate //////
+    var min = this.AABB.minimum;
+    var max = this.AABB.maximum;
+    var center = this.AABB.center;
+    min.x = Number.POSITIVE_INFINITY;
+    min.y = Number.POSITIVE_INFINITY;
+    min.z = Number.POSITIVE_INFINITY;
+    max.x = Number.NEGATIVE_INFINITY;
+    max.y = Number.NEGATIVE_INFINITY;
+    max.z = Number.NEGATIVE_INFINITY;
+
+    var resolution = this.resolution;
+
+    // Find the min/max bounds and resolution of the uniform grid
+    var objectCount = data.length;
+    for (var i = 0; i < objectCount; i++) {
+        objectAABBFunction(data[i], min, max);
+    }
+
+    // Figure out what the grid's resolution should be. Pad min and max out to match
+    resolution.x = Math.floor((max.x - min.x) / cellWidth) + 1;
+    resolution.y = Math.floor((max.y - min.y) / cellWidth) + 1;
+    resolution.z = Math.floor((max.z - min.z) / cellWidth) + 1;
+
+    center = Cartesian3.add(min, max, center);
+    center = Cartesian3.divideByScalar(center, 2.0, center);
+    min.x = center.x - (resolution.x / 2.0) * cellWidth;
+    min.y = center.y - (resolution.y / 2.0) * cellWidth;
+    min.z = center.z - (resolution.z / 2.0) * cellWidth;
+    max.x = center.x + (resolution.x / 2.0) * cellWidth;
+    max.y = center.y + (resolution.y / 2.0) * cellWidth;
+    max.z = center.z + (resolution.z / 2.0) * cellWidth;
+
+    var cellCount = resolution.x * resolution.y * resolution.z;
+
+    // Bin the objects
+    var allCellData = [];
+    for (i = 0; i < cellCount; i++) {
+        allCellData.push({
```

Comment:
The objects in this array are used a lot, right?  Instead of an object literal, make a constructor function for a new type for better performance.

See https://www.youtube.com/watch?v=XAqIpGU8ZZk